Enhance: make multiView persistent

SlySven opened the pull request:
~~Makes the compact input line a toggle (so now it also has constant text).~~

Makes the multi-view stay active when `<Ctrl>+<Tab>` / `<Ctrl>+Number` switching between - or clicking on the tab bars.

Also disables the (now a toggle) control for multi-view when there is less than two profiles active.

This should close #1111 #1831 #3136 #3204 ~~#2400~~, it may also improve #2099

~~It also disables dragging the tabs around so will prevent the issue mentioned in #1456 pending a fix that can also rearrange the main `TConsole`s to match.~~

Also improve layout of the Multi-view tool-tip for both the toolbar and menu instances of the control.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: make multiView persistent & compact input line act on all profiles

Makes the compact input line a toggle (so now it also has constant text).

Makes the multi-view stay active when `<Ctrl>+<Tab>` / `<Ctrl>+Number`
switching between - or clicking on the tab bars.

Also disables the (now a toggle) control for multi-view when there is less
than two profiles active.

This should close #1111 #1831 #3136 #3204 #2400, it may also improve #2099

It also disables dragging the tabs around so will prevent the issue
mentioned in #1456 pending a fix that can also rearrange the main
`TConsole`s to match.

Also improve layout of the Multi-view tool-tip for both the toolbar and
menu instances of the control.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented on src/mudlet.cpp:
```diff
@@ -407,8 +411,11 @@ mudlet::mudlet()
 
     mpActionMultiView = new QAction(QIcon(QStringLiteral(":/icons/view-split-left-right.png")), tr("MultiView"), this);
-    mpActionMultiView->setToolTip(tr("If you've got multiple profiles open, splits Mudlet screen to show them all at once"));
+    mpActionMultiView->setToolTip(tr("<p>Enabled when multiple profiles are open, this splits the Mudlet screen to show them all at once.</p>"));
```
The text seems a bit convoluted now. Split into two sentences.
Do not give HTML tags to translators unless utterly necessary.

```suggestion
    mpActionMultiView->setToolTip(QString("<p>%1</p>").arg(tr("Splits the Mudlet screen to show multiple profiles at once. Deactivated while you open less.")));
```
Code has not been tested, just the general idea here.

SlySven replied:
You didn't use `QStringLiteral(`...`)` so that will still be marked as an ambiguous `QString` - there is a `(static QString) mudlet::htmlWrapper` defined currently as `QStringLiteral("<html><head /><body>%1</body></html>")` which I put in a while back to cover this exact situation. However, since then I found it should be trimmed to be `QStringLiteral("<p>%1</p>")` - but then it would mean some other unrelated tooltips would have to have existing `<p>`...`</p>` taken out from their Engineering English text - thus making this PR even larger...

I was also in two minds as whether it is cleaner to leave the bracketing `<p>`...`</p>` in place so the in the odd occasion where more than one paragraph is wanted - to separate, perhaps a significant detail - the translator is not presented with an odd looking `</p><p>` in the **middle** of a text *without* the matching tags at either end - which have been concealed from them - even if it doesn't confuse them I am not sure that the CrowdIn auto-hiding of tags is going to make sense of those...

SlySven replied:
`"Splits the Mudlet screen to show multiple profiles at once; disabled when less than two are loaded."`?

"Deactivated while you open less." is not grammatically correct IMHO - though I do not know if I can explain why...

Also I am working on another PR that could get the user back to the state as if they have cancelled the initial "Open a profile" dialogue so increasing the possibility of not having ANY profiles loaded/connected...

Kebap replied:
Fair enough. Not sure why you prefer semicolon to seperate sentences but meh.

SlySven replied:
Dammit - I've been rumbled; semicolons enable me to keep going for longer without having to pause or otherwise interrupt my narrative by coming to a halt, period (or, if you like) a full stop **.**

Kebap commented on src/mudlet.cpp:
```diff
@@ -1623,6 +1630,26 @@ void mudlet::slot_tab_changed(int tabID)
     // update the window title for the currently selected profile
     setWindowTitle(mpCurrentActiveHost->getName() + " - " + version);
 
+    // Restore the multi-view mode if it enabled:
```
```suggestion
    // Restore the multi-view mode if it was enabled:
```

Kebap commented on src/mudlet.cpp:
```diff
@@ -4236,49 +4262,58 @@ void mudlet::slot_connection_dlg_finished(const QString& profile, bool connect)
 }
 
-void mudlet::slot_multi_view()
+// Needed because the use of a shortcut to trigger the menuaction does not
+// provide the checked state of the item to which the shortcut is associated:
```
Comment hardly understandable. 
2 line comment for 2 line code.. 🤷‍♂️

SlySven replied:
The compilation failure when I tried to connect a signal with NO arguments to a slot that required ONE was even more not understandable!!! It took me a bit of researching to find a solution - especially as I wasn't the one who put in the `QShortcuts`...

Kebap commented on src/mudlet.cpp:
```diff
@@ -4236,49 +4262,58 @@ void mudlet::slot_connection_dlg_finished(const QString& profile, bool connect)
 }
 
-
+// Called by short-cut that doesn't pass the menu-items checked state
```
Am I alone in not understanding these? 🤔

vadi2 replied:
Not alone, I don't get it too

SlySven replied:
The signal created by the use of say the `(QShortcut) mudlet::multiViewShortcut` does not carry the `Qt::Checked`/`Qt::Unchecked` argument that the - now *checkable* `(QAction*) mudlet::dactionMultiView` {a.k.a. a `QMenuItem`} and `(QAction*) mudlet::dactionMultiView`  {a.k.a. a `QToolButton`} produce as an argument of their `QAction::triggered` signal.

So: I had to provide a second slot {`(void) mudlet::slot_toggle_multi_view()`} to connect it to that considered the state of the new `(bool) mudlet::mMultiView` flag - toggled it and then passed that value to the existing `(void) mudlet::slot_multi_view(const bool)` which has been extended you might note to record the checked state of the control that generated the signal (and which ensures that both controls reflect that state as part of its execution).

\<aside>Having done that for the multi-view function - i.e. making it a toggle - it seemed an eminently suitable way to solve the existing "Standard/Compact" input line problem - which came about as @vadi2 coded it act in a global way (by maintaining the state in a `mudlet` class member) but only applied it to **one** profile - completely forgetting about multi-playing IMVHO.\</aside>

Kebap commented on src/mudlet.h:
```diff
@@ -723,6 +725,9 @@ private slots:
     // const TBuffer::csmEncodingTable:
     QMap<QByteArray, QString> mEncodingNameMap;
+
+    // Whether multi-view is in effect (if more than one tab is present)
```
Not sure if this is the right place to explain multi-view elementals
```suggestion
    // Whether multi-view is in effect
```

SlySven replied:
Explaining what a flag does/controls/indicates? This is one of the best places for it as in a search you can always pick out (and in fact IDEs include the Qt one home in on) the place something is ***declared***.

Kebap replied:
Yes `mMultiView` indicates "Whether multi-view is in effect"
That's fair, that is my suggestion as well.
Whereas "Multi-view is only possible if more than one tab is present" - has nothing to do with mMultiView.

SlySven replied:
:thinking: *well* I sort of see what you are saying - and in practice this flag will be reflected in the up/down state of the toolbar button and the unchecked/checked state of the menu item. Whether those controls are *enabled* is a separate issue (which is tied to the count of tabs present as it happens, now!) and if multi-view was present when the user gets back to only having one profile open, the controls will stay (locked) in their up/down/unchecked/checked state and will not be alterable until a second profile is active again...

Kebap commented on src/ui/main_window.ui:
```diff
@@ -289,16 +289,31 @@
+   </property>
    <property name="toolTip">
-    <string>&lt;p&gt;Share the screen area between multiple open profiles.&lt;/p&gt;</string>
+    <string>&lt;p&gt;Enabled when multiple profiles are open, this splits the Mudlet screen to show them all at once.&lt;/p&gt;</string>
```
Same text for button and menu action. :+1:
💡 Maybe move all menu tooltips to .cpp in the future to avoid HTML for translators.. 
```suggestion
    <string>&lt;p&gt;Splits the Mudlet screen to show multiple profiles at once. Deactivated while you open less.&lt;/p&gt;</string>
```

SlySven replied:
> :bulb: Maybe move all menu tooltips to .cpp in the future to avoid HTML for translators..

I have considered that on a more general basis - then you end up with `dlgAboutDialog.cpp` :grinning: .

Actually though it isn't necessary - `lupdate` gathers identical texts from all sources  (and despite the use of HTML entities the texts are treated as the same) and presents them as a single text for translation - provided that the coders remember that there are multiple instances - of course the comments also have to be identical so individually numbering them (e.g. "x of y") in those comments stops that working - whereas ("ensure all y copies are kept the same") can alert others to the issue.

Kebap replied:
Pretty sure that only works for strings in a same file though. 
So no dice for .ui files doubling .cpp texts like in this case.

SlySven replied:
> Pretty sure that only works for strings in a same file though.

Nope - I'm pretty sure that it works for all files in the application - it is some of the black magic that `lupdate` does and why it is not easy to switch away from it - as it knows how to look in `.ui` files as well as `.h` and `.cpp` ones... - oh, looking at the `mudlet.ts` file I see it does separate the string sources by *classname* OTOH the Main window file might have been in the same class - if it had been called, say `mudlet.ui` rather than `main_window.ui` :man_shrugging:

SlySven replied:
Ah, well, at least the translation memory system will spot identical source texts even in different files and suggest the translation of the first one in subsequent duplicate source texts in different files...

Kebap reviewed (approved):
I for one welcome this nice effort to close a lot of long-standing problems! 👍 

Then again, there are a few distinct changes mixed in a same PR.
Some of them I found debatable, and will elaborate below.
Others could and should better make it to 4.9 asap..!

New features work as expected here on Windows as well:
- Ctrl+Tab switches profiles, and keeps multi-view intact
- Alt+L toggles compact toolbar on all profiles at once
- Dragging profile tabs around isn't working anymore

> Makes the multi-view stay active when `<Ctrl>+<Tab>` / `<Ctrl>+Number` switching between - or clicking on the tab bars.

Only minor niggle... The two ways you name will change the focus of the profiles' tab. Nice!
However, when I left-click IN the other profile, its text entry box for example will gain the cursor.
The old profile seems like it's still active (as its tab still has focus) whereas I clearly changed profile.
(It was also like this in current version, and no bug report mentioned it, so maybe out of scope here)
Another test as well: After clicking in the new profile, Alt+E will open the script editor for the old profile.

> Makes the compact input line a toggle (so now it also has constant text).

Maybe better if this stayed profile-specific...?
As it stands, now either all or no profile have it.
I saw no way to toggle only one profile compact.
The current Mudlet version lets you toggle differently.
So this actually feels like a regression! Maybe unintended.

> It also disables dragging the tabs around so will prevent the issue mentioned in #1456 pending a fix that can also rearrange the main `TConsole`s to match.

Not sure if I like this "solution" to the reported issue. 
I'd rather have the fix. We'll leave that issue open, no?
Then we can merge now and improve more in the future.

vadi2 commented:
I've never multiplayed so I'll defer the decision-making to the people who do.

Kebap commented:
Persistent multi-view has been requested often by many players as you can review in multiple issues linked at the top. I'd love to see this part of the PR come alive asap. My feature requests here can be postponed if they get in the way. 

However, have you ever had more than one profile open simultaneously? No need for multi-view enabled. Already without that, the compact input line regression may affect you as well. I am not sure if we're smart to ship it like this. 

That is why I keep promoting granular PRs.

vadi2 commented:
We definitely shouldn't - I'd say you should rescind your approval.

Kebap reviewed (changes requested):
Make compact input profile specific or exclude from this PR for further discussion

SlySven commented:
> > It also disables dragging the tabs around so will prevent the issue mentioned in #1456 pending a fix that can also rearrange the main TConsoles to match.

> Not sure if I like this "solution" to the reported issue.

I ran into this issue whilst working in this area - and it is a noticeably bad thing from a UI point of view - you drag the tabs around and they move freely enough and settle in place - but then they are totally unrelated to the `TConsole` they sit on top of and you can end up closing the wrong profile. This probably came about because someone wasn't aware of the (default ?) movability of tabs on a tab bar. So as to stop this breakage I felt it was sensible to lock the tabs above their profiles ***until*** we can find a way to sort them out - I am not entirely sure how to do it but I have ideas. Anyhow, until this gets addressed, this step will prevent user confusion which has to be a good thing?

SlySven reviewed (commented):
Please consider my feedback to your feedback...

SlySven commented:
Also:
> Make compact input profile specific or exclude from this PR for further discussion

The fix to make it global is quicker and easier than doing it on a per profile basis:
For starters you'd need to move the state flag to the `Host` class - and save and load it - so changes to `XMLimport` and `XMLexport` classes, then, whether you kept my use of a checkable option (with no changing of the text BTW) or reverted to the changing text, you have to ensure the indication of the state was changed each time the active profile changed.

All in all, a global change did not seem that unreasonable.  Besides - if you are multi-playing - you will probably want as much space for your individual input lines as possible then - and having to switch them off individually might actually be less convenient than doing them all at once...

vadi2 commented:
I don't think it's a bug to be honest and I don't think anyone is confused - rather it was a wishlist item to improve the behaviour...

SlySven commented:
Sorry - which bit?

vadi2 commented:
The shuffling of the input tabs.

SlySven commented:
As I said I hadn't really noticed it before but having run into it whilst working in this area this is a temporary :adhesive_bandage: amelioration which will stop the user being misled like I was - it does not reduce the functionality but instead reinforces the current state of affairs - the profiles' order can *not* be rearranged - yet!

SlySven commented:
:thinking:  Ah, you might be referring to something else after all - let me review things and check that all the things apply to the correct idea of what is the *currently Active* profile...

Kebap commented:
Damn now we are discussing 3 items all jumbled together, just like I predicted.

Kebap commented:
I feel we focus on the wrong items. 

How do we decide if we merge that input line regression alongside the multi-view fix or no? 

Do you see a chance as well to include changing profile focus when clicked upon?

vadi2 commented:
We shouldn't merge with a regression, that's for sure.

SlySven committed:
Merge branch 'development' into Enhance_makeMultiViewPersistent&CompactInputLineActOnAllProfiles

Conflicts resolved in:
* src/TConsole.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revoke: remove changes disassociated with the multi-view issues

Required to simplify this PR and so they can be moved to a separate
PR.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Okay I have taken out the stuff related to the compact input line (and the locking the tabs to the profiles change) and moved them to separate PRs. I will now look thought the comments and consider the other points that were raised...

Kebap reviewed (approved):
Sombunal comments improved as well! 👍

SlySven committed:
Enhance: improve active tab indication in multi-view

Examining the situations in which clicking on a part of the UI does not
cause the indication of the active tab to reflect the profile that the
widget is part of has shown a few places which are addressed in this
commit:
* 2D Mapper
* 3D Mapper
* Search term entry widget in the commandl ine
* either `TTextEdit` in the Main or User Window `TConsole`
* `TLabels` in any `TConsole`

All of the above will now call a new mudlet::activateProfile(Host*) which
will ensure that the correct tab is selected, and that the member:
`(Host*) mudlet::mpCurrentActiveHost` is updated to also reflect that.

Also, if multi-view is active than there is not need to highlight the
tabs of profiles that have seen any changes - which is just as well as the
code to clear the tab is dependent on the user clicking on the affected
tab to show the tab and clear it - which is less likely to happen when in
the multi-view mode.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_makeMultiViewPersistent&CompactInputLineActOnAllProfiles

Needed to potentially get access to some new functions in the
TTabBar class.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
I have tried to address the issue of clicking on widgets for a profile that is not the *current* one and have a solution that should cause the selected tab when multi-playing to follow the currently active profile better.

SlySven commented:
Refresh me - is there anything that is still objectionable about this PR that now only focuses on the multi-view aspect?

itsTheFae reviewed (approved):
I see no issues with this.  The test build is working as advertised. :)

SlySven commented:
Thanks for approving this @itsTheFae - sadly it looks like in the rush to get it in, the commit message for the squash and merge did not get edited to reflect what it actually now does - so one might be mislead if one only reads the first few lines...
